modules/ShowersMatcher.py

```diff
@@ -0,0 +1,427 @@
+import networkx as nx
+import numpy as np
+import matplotlib.pyplot as plt
+import pandas as pd
+import tensorflow as tf
+from scipy.optimize import linear_sum_assignment
+
+def calculate_iou_tf(truth_sid,
+                     pred_sid,
+                     truth_shower_sid,
+                     pred_shower_sid,
+                     hit_weight, return_all=False):
+
+    with tf.device('/cpu:0'):
+        # print("1")
+        truth_sid = tf.cast(tf.convert_to_tensor(truth_sid), tf.int32)
+        pred_sid = tf.cast(tf.convert_to_tensor(pred_sid), tf.int32)
+        hit_weight = tf.cast(tf.convert_to_tensor(hit_weight), tf.float32)
+        # print("2")
+
+        truth_shower_sid = tf.cast(tf.convert_to_tensor(truth_shower_sid), tf.int32)
+        pred_shower_sid = tf.cast(tf.convert_to_tensor(pred_shower_sid), tf.int32)
+        len_pred_showers = len(pred_shower_sid)
+        len_truth_showers = len(truth_shower_sid)
+
+        # print("3")
+        truth_idx_2 = tf.zeros_like(truth_sid) -1
+        pred_idx_2 = tf.zeros_like(pred_sid) -1
+        hit_weight_2 = tf.zeros_like(hit_weight)
+
+        # print("3.1")
+
+        for i in range(len(pred_shower_sid)):
+            pred_idx_2 = tf.where(pred_sid == pred_shower_sid[i], i, pred_idx_2)
+
+        for i in range(len(truth_shower_sid)):
+            truth_idx_2 = tf.where(truth_sid == truth_shower_sid[i], i, truth_idx_2)
+
+        # print("3.3")
+        one_hot_pred = tf.one_hot(pred_idx_2, depth=len_pred_showers)
+        one_hot_truth = tf.one_hot(truth_idx_2, depth=len_truth_showers)
+
+        intersection_sum_matrix = tf.linalg.matmul(one_hot_pred * hit_weight[..., tf.newaxis], one_hot_truth, transpose_a=True)
+
+        pred_sum_matrix = tf.linalg.matmul(one_hot_pred * hit_weight[..., tf.newaxis], tf.ones_like(one_hot_truth),
+                                        transpose_a=True)
+
+        truth_sum_matrix = tf.linalg.matmul(
+            tf.ones_like(one_hot_pred) * hit_weight[..., tf.newaxis], one_hot_truth, transpose_a=True)
+
+        union_sum_matrix = pred_sum_matrix + truth_sum_matrix - intersection_sum_matrix
+
+        overlap_matrix = (intersection_sum_matrix / union_sum_matrix).numpy()
+
+        if return_all:
+            return overlap_matrix, pred_sum_matrix, truth_sum_matrix, intersection_sum_matrix
+        else:
+            return overlap_matrix
+
+
+def angle(p, t):
+    t = np.array([t['x'], t['y'], t['z']])
+    p = np.array([p['dep_x'], p['dep_y'], p['dep_z']])
+
+    angle = np.arccos(np.sum(t*p) / (np.sqrt(np.sum(t*t))*np.sqrt(np.sum(p*p))))
+
+    return angle
+
+
+class ShowersMatcher:
+    _NODE_TYPE_TRUTH_SHOWER = 0
+    _NODE_TYPE_PRED_SHOWER = 1
+    _NODE_TYPE_RECHIT = 7
+
+    def __init__(self, match_mode, iou_threshold, de_e_cut, angle_cut):
+        super().__init__()
+        self.match_mode=match_mode
+        self.iou_threshold=iou_threshold
+        self.de_e_cut=de_e_cut
+        self.angle_cut=angle_cut
```

Comment:
a quick check, e.g.
assert match_mode in ['eiou', ....]
Then it is clear why it fails if a wrong option is given